### PR TITLE
Implement vector:getAngle() with lua functions.

### PR DESCRIPTION
**Pull request by ANormalTwig**
Use Lua functions to turn a vector into an angle instead of unwrapping and using the normal GLua ones.

### CODE REVIEW

**Comment by ANormalTwig** [comment]
I don't know if the slight difference is any concern but the returned angle is slightly different since lua uses doubles instead of floats like the source engine angles.

**Comment by thegrb93** [comment]
Need a performance comparison

**Comment by ANormalTwig** [comment]
Don't have any good reason, just figured it'd be faster with the lack of unwrapping.

image
image

**Comment by thegrb93** [comment]
Could you run that test one more time and make sure the angle results of the before and after are always close?

**Comment by ANormalTwig** [comment]
There is basically no difference the new one just has more decimal points.
image
image
image
image

File: lua/starfall/libs_sh/vectors.lua [hunk 1]
```diff
@@ -189,10 +189,17 @@ function vec_meta.__eq(a, b)
...
 end
 
+local math_asin, math_atan2, math_sqrt = math.asin, math.atan2, math.sqrt
+local rad2deg = 180 / math.pi
+
 --- Get the vector's angle.
 -- @return Angle Angle representing the vector
 function vec_methods:getAngle()
-	return awrap(unwrap(self):Angle())
+	return setmetatable({
```
Comment (thegrb93) [inline lua/starfall/libs_sh/vectors.lua:198]: I think the `%360` can be removed. Also if the vector length `math_sqrt(self[1]^2 + self[2]^2 + self[3]^2)` is zero, we should return a zero angle. Currently, it returns an angle with nan pitch.

Can you do the performance test again after that change?

**Review by thegrb93 — CHANGES_REQUESTED** [review]

File: lua/starfall/libs_sh/vectors.lua (hunk 1 above)
Comment (ANormalTwig) [reply lua/starfall/libs_sh/vectors.lua:198]: the % 360 is for parity with the original since the original never returns an angle with any negatives

**Comment by thegrb93** [comment]
Can you do also do your comparison again with `Vector(math.random()-0.5, math.random()-0.5, math.random()-0.5)` so we get some negative vectors as well?

**Comment by ANormalTwig** [comment]
image
performance after zero checking the length

image
largest diffrence when including negative vectors

**Pull request merged**